### PR TITLE
Broker Relist

### PR DESCRIPTION
**Pull request by eriknelson**
Initial fields for #1086

### CODE REVIEW

File: pkg/apis/servicecatalog/types.go [hunk 1]
```diff
@@ -60,8 +60,29 @@ type ServiceBrokerSpec struct {
 	// CABundle is a PEM encoded CA bundle which will be used to validate a Broker's serving certificate.
 	// +optional
 	CABundle []byte
+
+	// RelistBehavior specifies the type of relist behavior the catalog should exhibit
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: We need to add validations for these fields.  See pkg/apis/servicecatalog/validation/broker.go

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 2]
```diff
@@ -60,8 +60,29 @@ type ServiceBrokerSpec struct {
 	// CABundle is a PEM encoded CA bundle which will be used to validate a Broker's serving certificate.
 	// +optional
 	CABundle []byte `json:"caBundle,omitempty"`
+
+	// RelistBehavior specifies the type of relist behavior the catalog should exhibit
+	// when relisting ServiceClasses available from a broker
+	RelistBehavior ServiceBrokerRelistBehavior
```
Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: fields in versioned APIs need to have json tags associated with them.  For this field, it would be:

```go
RelistBehavior ServiceBrokerRelistBehavior `json:"relistBehavior"`
```

You'll need to add these to the fields in this file.

File: pkg/controller/controller_broker.go [hunk 3]
```diff
@@ -113,8 +113,8 @@ const (
 // shouldReconcileServiceBroker determines whether a broker should be reconciled; it
 // returns true unless the broker has a ready condition with status true and
 // the controller's broker relist interval has not elapsed since the broker's
-// ready condition became true.
-func shouldReconcileServiceBroker(broker *v1alpha1.ServiceBroker, now time.Time, relistInterval time.Duration) bool {
+// ready condition became true, or if the broker's RelistBehavior is set to Manual
```
Comment (pmorie) [inline pkg/controller/controller_broker.go]: missing period

File: pkg/controller/controller_broker.go [hunk 4]
```diff
@@ -113,8 +113,8 @@ const (
...
+func shouldReconcileServiceBroker(broker *v1alpha1.ServiceBroker, now time.Time, relistInterval metav1.Duration) bool {
```
Comment (pmorie) [inline pkg/controller/controller_broker.go]: Technically you do not need to pass in the `relistInterval` parameter here, since you're now reading it from the broker.

File: pkg/controller/controller_broker.go [hunk 5]
```diff
@@ -168,14 +185,16 @@ func (c *controller) reconcileServiceBrokerKey(key string) error {
 // processed and should be resubmitted at a later time.
 func (c *controller) reconcileServiceBroker(broker *v1alpha1.ServiceBroker) error {
 	glog.V(4).Infof("Processing ServiceBroker %v", broker.Name)
+	if broker.Spec.RelistDuration == nil {
```
Comment (pmorie) [inline pkg/controller/controller_broker.go]: I think this TODO is fixed :)

You should be defensive in programming, but some assumptions that you have validated for already are okay.  So for instance, you should ensure that you don't dereference a nil pointer for an optional field.  However, if that optional field must be set based on a validation, a state where that validation fails in code should be handled like an error.

File: pkg/controller/controller_broker.go [hunk 6]
```diff
@@ -168,14 +185,16 @@ func (c *controller) reconcileServiceBrokerKey(key string) error {
...
+	// * If the broker's ready condition is true and the RelistBehavior has been
+	// set to Manual, do not reconcile it.
+	// * If the broker's ready condition is true and the relist interval has not
 	// elapsed, do not reconcile it.
-	if !shouldReconcileServiceBroker(broker, time.Now(), c.brokerRelistInterval) {
-		glog.V(10).Infof(
-			"Not processing ServiceBroker %v because relist interval has not elapsed since the broker became ready",
-			broker.Name,
-		)
+	if !shouldReconcileServiceBroker(broker, time.Now(), *broker.Spec.RelistDuration) {
```
Comment (pmorie) [inline pkg/controller/controller_broker.go]: Your TODO appears to be handled now in `shouldReconcileServiceBroker`.

File: pkg/controller/controller_broker.go (hunk 4 above)
Comment (pmorie) [reply pkg/controller/controller_broker.go]: Oh, I didn't realize you were still reading it as a parameter.  You already have the broker here - you can read the value in this method and delete the `relistInterval` parameter.  There needs to ve a validation that ensures that the duration field is set when the mode is 'duration'.

File: pkg/controller/controller_broker_test.go [hunk 7]
```diff
@@ -553,6 +554,8 @@ func testReconcileServiceBrokerWithAuth(t *testing.T, authInfo *v1alpha1.Service
 	}
 }
 
+// ERIK:TODO: Write up another test here to confirm an error is bubbled when
+// RelistDuration is nil, similar to this?
```
Comment (pmorie) [inline pkg/controller/controller_broker_test.go]: RelistDuration being nil should be handled in `shouldReconcileBroker`.

File: pkg/storage/tpr/storage_interface_test.go [hunk 8]
```diff
@@ -140,7 +141,9 @@ func TestCreateWithNoNamespace(t *testing.T) {
 	inputServiceBroker := &sc.ServiceBroker{
 		ObjectMeta: metav1.ObjectMeta{Name: name},
 		Spec: sc.ServiceBrokerSpec{
-			URL: "http://my-awesome-broker.io",
+			URL:            "http://my-awesome-broker.io",
```
Comment (pmorie) [inline pkg/storage/tpr/storage_interface_test.go]: I am not qualified to review TPR stuff.  @arschles @nilebox

File: pkg/apis/servicecatalog/testing/fuzzer.go [hunk 9]
```diff
@@ -192,6 +193,11 @@ func FuzzerFor(t *testing.T, version schema.GroupVersion, src rand.Source) *fuzz
 			// Set the bytes field on the RawExtension
 			r.Raw = bytes
 		},
+		func(bs *servicecatalog.ServiceBrokerSpec, c fuzz.Continue) {
```
Comment (pmorie) [inline pkg/apis/servicecatalog/testing/fuzzer.go]: I'm surprised that the default fuzzer didn't work for this field.  Do you try it?

File: pkg/apis/servicecatalog/v1alpha1/defaults.go [hunk 10]
```diff
@@ -18,13 +18,25 @@ package v1alpha1
...
 func addDefaultingFuncs(scheme *runtime.Scheme) error {
 	return RegisterDefaults(scheme)
 }
 
+func SetDefaults_ServiceBrokerSpec(spec *ServiceBrokerSpec) {
+	if spec.RelistBehavior == "" {
+		spec.RelistBehavior = ServiceBrokerRelistBehaviorDuration
+	}
+
+	if spec.RelistDuration == nil {
```
Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/defaults.go]: You should only default relist duration if the relist behavior is "Duration".  It should not be defaulted for manual behavior.

File: pkg/apis/servicecatalog/v1alpha1/defaults_test.go [hunk 11]
```diff
@@ -68,6 +68,22 @@ func roundTrip(t *testing.T, obj runtime.Object) runtime.Object {
 	return obj3
 }
 
+func TestSetDefaultServiceBroker(t *testing.T) {
```
Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/defaults_test.go]: Tests here need to change according to the logic as I commented above

Comment (pmorie) [reply pkg/apis/servicecatalog/v1alpha1/defaults_test.go]: There should be test cases for:

1.  Neither duration or behavior set
2.  Behavior set to manual
3.  Behavior set to duration but no duration provided

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 12]
```diff
@@ -60,8 +60,29 @@ type ServiceBrokerSpec struct {
 	// CABundle is a PEM encoded CA bundle which will be used to validate a Broker's serving certificate.
 	// +optional
 	CABundle []byte `json:"caBundle,omitempty"`
+
+	// RelistBehavior specifies the type of relist behavior the catalog should exhibit
+	// when relisting ServiceClasses available from a broker
+	RelistBehavior ServiceBrokerRelistBehavior `json:"relistBehavior"`
```
Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: newline separate these to keep it consistent with the rest of the file, same for the other types.go

File: pkg/apis/servicecatalog/validation/broker.go [hunk 13]
```diff
@@ -99,6 +99,12 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
 			)
 		}
 
+		if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration && spec.RelistDuration == nil {
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker.go]: Question: do you think it is valid for duration to be set if the behavior is manual?  It might be.  If you switch from duration to manual temporarily, for example, you might not expect to have to unset the duration.  Consider what the 'least surprise' is here.

Comment (pmorie) [reply pkg/apis/servicecatalog/validation/broker.go]: Also, I don't see unit tests for the validations

File: pkg/controller/controller_broker_test.go [hunk 14]
```diff
@@ -54,77 +54,82 @@ func TestShouldReconcileServiceBroker(t *testing.T) {
 	// name: short description of the test
 	// broker: broker object to test
 	// now: what time the interval is calculated with respect to interval
-	// internal: the time that has elapsed since now
 	// reconcile: whether or not the reconciler should run, the return of
 	// shouldReconcileServiceBroker
+	// ERIK:TODO: Test new Relist codepaths
```
Comment (pmorie) [inline pkg/controller/controller_broker_test.go]: Looks like this TODO still needs to be done, accurate?

File: pkg/apis/servicecatalog/testing/fuzzer.go (hunk 9 above)
Comment (eriknelson) [reply pkg/apis/servicecatalog/testing/fuzzer.go]: Think I did, but I'll pull it and try again.

File: pkg/controller/controller_broker_test.go (hunk 14 above)
Comment (eriknelson) [reply pkg/controller/controller_broker_test.go]: Have a local commit with a bunch of error handling I wanted to run by you. It tests the RelistDuration == nil path, but I want to make sure I hit the rest of them.

File: pkg/apis/servicecatalog/validation/broker.go (hunk 13 above)
Comment (eriknelson) [reply pkg/apis/servicecatalog/validation/broker.go]: > do you think it is valid for duration to be set if the behavior is manual? It might be. If you switch from duration to manual temporarily, for example, you might not expect to have to unset the duration.

Agreed. I would be surprised if I had to explicitly unset duration, and instead would expect the duration to simply be ignored.

File: pkg/apis/servicecatalog/types.go [hunk 15]
```diff
@@ -60,8 +60,31 @@ type ServiceBrokerSpec struct {
...
+	// when relisting ServiceClasses available from a broker
+	RelistBehavior ServiceBrokerRelistBehavior
+
+	//RelistDuration is the frequency by which a controller will relist the broker
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: There should be spaces after the `//` and all godoc comments for fields should end in periods.

File: pkg/apis/servicecatalog/types.go [hunk 16]
```diff
@@ -60,8 +60,31 @@ type ServiceBrokerSpec struct {
...
+	//RelistRequests is a strictly increasing integer counter that can be manually incremented
+	//by a user to manually trigger a relist
+	RelistRequests int64
 }
 
+// ServiceBrokerRelistBehavior represents a type of broker relist behavior
+type ServiceBrokerRelistBehavior string
+
+const (
+	// ServiceBrokerRelistBehaviorDuration will allow the RelistDuration frequency to be configured
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: `// ServiceBrokerRelistBehaviorDuration indicates that the broker will be relisted automatically after the specified duration has passed.`

File: pkg/apis/servicecatalog/types.go [hunk 17]
```diff
@@ -60,8 +60,31 @@ type ServiceBrokerSpec struct {
...
+	ServiceBrokerRelistBehaviorDuration ServiceBrokerRelistBehavior = "Duration"
+	// ServiceBrokerRelistBehaviorManual will configure the catalog to only perform future
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: `// ServiceBrokerRelistBehaviorManual indicates that the broker is only relisted when the spec of the broker changes.`

File: pkg/apis/servicecatalog/v1alpha1/defaults_test.go (hunk 11 above)
Comment (pmorie) [reply pkg/apis/servicecatalog/v1alpha1/defaults_test.go]: Still need test cases for these defaults.  You need to ensure that after an object starting in each of the states in the above comment is defaults, that the object is in the expected state.

File: pkg/apis/servicecatalog/validation/broker.go [hunk 18]
```diff
@@ -99,6 +99,12 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
 			)
 		}
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker.go]: Need a validation here to ensure that the relist behavior is set.

Comment (pmorie) [reply pkg/apis/servicecatalog/validation/broker.go]: Still need to update the unit tests for this validation.

File: pkg/controller/controller_broker.go [hunk 19]
```diff
@@ -133,10 +133,36 @@ func shouldReconcileServiceBroker(broker *v1alpha1.ServiceBroker, now time.Time,
...
+					// ignore the Duration and only relist based on spec changes
+					glog.V(10).Infof(
+						"Not processing ServiceBroker %v: RelistBehavior is set to Manual",
+						broker.Name,
+					)
+					return false
+				}
+
+				if broker.Spec.RelistDuration == nil {
+					glog.V(10).Infof(
```
Comment (pmorie) [inline pkg/controller/controller_broker.go]: This should be logged at Error

File: pkg/apis/servicecatalog/validation/broker_test.go [hunk 20]
```diff
@@ -208,12 +229,56 @@ func TestValidateServiceBroker(t *testing.T) {
...
+					URL:            "http://example.com",
+					CABundle:       []byte("fake CABundle"),
+					RelistBehavior: "Duration",
+					RelistDuration: &metav1.Duration{15 & time.Minute},
+				},
+			},
+			valid: true,
+		},
+		{
+			name: "valid broker - manual behavior with behavior",
```
Comment (jboyd01) [inline pkg/apis/servicecatalog/validation/broker_test.go]: nit: name should be manual behavior with **duration**   instead of manual behavior with beavior

File: pkg/apis/servicecatalog/validation/broker.go [hunk 21]
```diff
@@ -98,13 +98,25 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
-
 	}
 
 	if spec.InsecureSkipTLSVerify && len(spec.CABundle) > 0 {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("caBundle"), spec.CABundle, "caBundle cannot be used when insecureSkipTLSVerify is true"))
 	}
 
+	if "" == spec.RelistBehavior {
+		allErrs = append(allErrs,
+			field.Required(fldPath.Child("relistDuration"),
```
Comment (staebler) [inline pkg/apis/servicecatalog/validation/broker.go]: The path here should be `relistBehavior` and the comment should indicate that the broker must have a relistBehavior.

File: pkg/apis/servicecatalog/validation/broker.go [hunk 22]
```diff
@@ -98,13 +98,25 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
+				"brokers must have a relistDuration"))
+	}
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration && spec.RelistDuration == nil {
+		allErrs = append(
+			allErrs,
+			field.Required(fldPath.Child("relistDuration"), "relistDuration must be set if behavior is set to Duration"),
```
Comment (staebler) [inline pkg/apis/servicecatalog/validation/broker.go]: This should probably be "relistDuration must be set if **relistBehavior** is set to Duration", so that it is clear which field is of importance.

File: pkg/apis/servicecatalog/validation/broker.go [hunk 23]
```diff
@@ -98,13 +98,25 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
+		)
+	}
+
```
Comment (staebler) [inline pkg/apis/servicecatalog/validation/broker.go]: Do we want to enforce that `relistRequests` be non-negative.

File: pkg/apis/servicecatalog/v1alpha1/defaults_test.go [hunk 24]
```diff
@@ -68,6 +70,64 @@ func roundTrip(t *testing.T, obj runtime.Object) runtime.Object {
...
+		actualSpec := ab.Spec
+
+		if tc.behavior != actualSpec.RelistBehavior {
+			t.Errorf(
+				"%v: unexpected default RelistBehavior: expected %v, got %v",
+				tc.name, tc.behavior, actualSpec.RelistBehavior,
+			)
+		}
+
+		if tc.duration == nil && actualSpec.RelistDuration == nil {
```
Comment (staebler) [inline pkg/apis/servicecatalog/v1alpha1/defaults_test.go]: This check does not verify that the actual RelistDuration was nil for the "behavior set to manual" test.

File: pkg/apis/servicecatalog/validation/broker.go [hunk 25]
```diff
@@ -98,13 +98,25 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("caBundle"), spec.CABundle, "caBundle cannot be used when insecureSkipTLSVerify is true"))
 	}
 
+	if "" == spec.RelistBehavior {
+		allErrs = append(allErrs,
+			field.Required(fldPath.Child("relistDuration"),
+				"brokers must have a relistDuration"))
+	}
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration && spec.RelistDuration == nil {
```
Comment (staebler) [inline pkg/apis/servicecatalog/validation/broker.go]: We need to validate that the RelistDuration is above a reasonable minimum, at least above 0.

File: pkg/apis/servicecatalog/validation/broker_test.go [hunk 26]
```diff
@@ -208,12 +229,56 @@ func TestValidateServiceBroker(t *testing.T) {
...
+					Name: "test-broker",
+				},
+				Spec: servicecatalog.ServiceBrokerSpec{
+					URL:            "http://example.com",
+					RelistBehavior: "Duration",
+					RelistDuration: nil,
+				},
+			},
+			valid: false,
+		},
```
Comment (staebler) [inline pkg/apis/servicecatalog/validation/broker_test.go]: We also need tests to verify the behavior when RelistBehavior is invalid and empty.

File: pkg/controller/controller_broker.go [hunk 27]
```diff
@@ -131,12 +131,39 @@ func shouldReconcileServiceBroker(broker *v1alpha1.ServiceBroker, now time.Time,
...
+					// ignore the Duration and only relist based on spec changes
+
+					glog.V(10).Infof(
+						"Not processing ServiceBroker %v: RelistBehavior is set to Manual",
+						broker.Name,
+					)
+					return false
+				}
+
+				if broker.Spec.RelistDuration == nil {
```
Comment (staebler) [inline pkg/controller/controller_broker.go]: Is this check necessary given that we have already performed validation on the Broker Spec to ensure that the RelistDuration is not nil when the RelistBehavior is not Manual?

File: pkg/apis/servicecatalog/types.go [hunk 28]
```diff
@@ -60,8 +60,31 @@ type ServiceBrokerSpec struct {
...
+
+	// RelistBehavior specifies the type of relist behavior the catalog should exhibit
+	// when relisting ServiceClasses available from a broker
+	RelistBehavior ServiceBrokerRelistBehavior
+
+	// RelistDuration is the frequency by which a controller will relist the broker
+	// when the RelistBehavior is set to ServiceBrokerRelistBehaviorDuration
+	RelistDuration *metav1.Duration
+
+	// RelistRequests is a strictly increasing integer counter that can be manually incremented
```
Comment (staebler) [inline pkg/apis/servicecatalog/types.go]: If we are claiming that this is a strictly increasing integer, then we should add a check in the ValidateUpdate function of the brokerRESTStrategy that validates that the new value is not less than the old value.

File: pkg/controller/controller_broker_test.go [hunk 29]
```diff
@@ -133,12 +138,33 @@ func TestShouldReconcileServiceBroker(t *testing.T) {
...
 				broker.Status.ReconciledGeneration = 1
+				broker.Spec.RelistDuration = &metav1.Duration{3 * time.Hour}
 				return broker
 			}(),
 			now:       time.Now(),
-			interval:  3 * time.Hour,
 			reconcile: true,
 		},
+		{
+			name: "ready, duration behavior, nil duration",
```
Comment (staebler) [inline pkg/controller/controller_broker_test.go]: We shouldn't need to test this. This should be caught by the Broker validator.

File: pkg/controller/controller_broker_test.go [hunk 30]
```diff
@@ -553,6 +585,7 @@ func testReconcileServiceBrokerWithAuth(t *testing.T, authInfo *v1alpha1.Service
 	}
 }
 
+// RelistDuration is nil, similar to this?
```
Comment (staebler) [inline pkg/controller/controller_broker_test.go]: Is this a stray comment that got placed here accidentally?

File: pkg/apis/servicecatalog/validation/broker.go (hunk 23 above)
Comment (eriknelson) [reply pkg/apis/servicecatalog/validation/broker.go]: I had a similar though. Should the field be a `uint`? I that normally acceptable?

File: pkg/controller/controller_broker.go (hunk 27 above)
Comment (eriknelson) [reply pkg/controller/controller_broker.go]: I've been back and forth on this. Are folks generally comfortable with trusting and dereffing a pointer value given a validation?

Comment (pmorie) [reply pkg/controller/controller_broker.go]: You should always program defensively, IMO.

File: pkg/apis/servicecatalog/v1alpha1/defaults_test.go (hunk 24 above)
Comment (eriknelson) [reply pkg/apis/servicecatalog/v1alpha1/defaults_test.go]: There is an open question as to whether or not RelistDuration MUST BE nil given a RelistBehavior of Manual. Is this something that needs to be asserted? CC @pmorie

File: pkg/controller/controller_broker_test.go (hunk 29 above)
Comment (eriknelson) [reply pkg/controller/controller_broker_test.go]: If it's being asserted in `shouldReconcileBroker`,  I think it's worth keeping as a test case. Happy to remove if folks feel strongly otherwise.

**Comment by eriknelson** [comment]
@staebler responded to feedback, thank you!

File: pkg/apis/servicecatalog/validation/broker.go (hunk 23 above)
Comment (staebler) [reply pkg/apis/servicecatalog/validation/broker.go]: `uint` is discouraged. All you need to do is add a validation check to ensure it is non-negative. And indicate that restriction in the comment.

Comment (eriknelson) [reply pkg/apis/servicecatalog/validation/broker.go]: @staebler pushed an update to the comments indicating as such. Validation asserts greater than zero.

File: pkg/apis/servicecatalog/validation/broker.go [hunk 31]
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
 	}
 
 	if spec.InsecureSkipTLSVerify && len(spec.CABundle) > 0 {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("caBundle"), spec.CABundle, "caBundle cannot be used when insecureSkipTLSVerify is true"))
 	}
 
+	if "" == spec.RelistBehavior {
+		allErrs = append(allErrs,
+			field.Required(fldPath.Child("relistBehavior"),
+				"brokers must have a relistBehavior"))
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker.go]: "relist behavior is required"

File: pkg/apis/servicecatalog/validation/broker.go [hunk 32]
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
+	}
+
+	isValidRelistBehavior := spec.RelistBehavior == "Duration" ||
+		spec.RelistBehavior == "Manual"
+	if !isValidRelistBehavior {
+		errMsg := fmt.Sprintf("unknown relistBehavior found: %v", spec.RelistBehavior)
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker.go]: "relist behavior must be "Manual" or "Duration""

File: pkg/apis/servicecatalog/validation/broker.go [hunk 33]
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("caBundle"), spec.CABundle, "caBundle cannot be used when insecureSkipTLSVerify is true"))
 	}
 
+	if "" == spec.RelistBehavior {
+		allErrs = append(allErrs,
+			field.Required(fldPath.Child("relistBehavior"),
+				"brokers must have a relistBehavior"))
+	}
+
+	isValidRelistBehavior := spec.RelistBehavior == "Duration" ||
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker.go]: use the constants from API types

File: pkg/apis/servicecatalog/validation/broker.go [hunk 34]
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
+	}
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration {
+		if spec.RelistDuration == nil {
+			allErrs = append(
+				allErrs,
+				field.Required(fldPath.Child("relistDuration"), "relistDuration must be set if behavior is set to Duration"),
+			)
+		}
+		if spec.RelistRequests < 0 {
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker.go]: I think you always want to apply this validation, even if the behavior is manual

File: pkg/apis/servicecatalog/validation/broker.go [hunk 35]
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration && spec.RelistDuration == nil {
+		allErrs = append(
+			allErrs,
+			field.Required(fldPath.Child("relistDuration"), "relistDuration must be set if relistBehavior is set to Duration"),
+		)
+	}
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration {
+		if spec.RelistDuration == nil {
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker.go]: looks like this is a duplicate of :124, and looks like this can be removed now, see comment about non-negative validation

File: pkg/apis/servicecatalog/validation/broker_test.go [hunk 36]
```diff
@@ -208,12 +229,82 @@ func TestValidateServiceBroker(t *testing.T) {
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-broker",
+				},
+				Spec: servicecatalog.ServiceBrokerSpec{
+					URL:            "http://example.com",
+					RelistBehavior: "",
+				},
+			},
+			valid: false,
+		},
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker_test.go]: need test for negative relistRequests

File: pkg/controller/controller_broker_test.go (hunk 29 above)
Comment (pmorie) [reply pkg/controller/controller_broker_test.go]: Keep it

File: pkg/apis/servicecatalog/v1alpha1/defaults.go [hunk 37]
```diff
@@ -18,13 +18,28 @@ package v1alpha1
...
 	"k8s.io/apimachinery/pkg/runtime"
+	"time"
 )
 
 func addDefaultingFuncs(scheme *runtime.Scheme) error {
 	return RegisterDefaults(scheme)
 }
 
+func SetDefaults_ServiceBrokerSpec(spec *ServiceBrokerSpec) {
+	isValidRelistBehavior := spec.RelistBehavior == "Duration" ||
```
Comment (nilebox) [inline pkg/apis/servicecatalog/v1alpha1/defaults.go]: Reference `ServiceBrokerRelistBehaviorDuration` and `ServiceBrokerRelistBehaviorManual` constants instead of inline strings?

File: pkg/apis/servicecatalog/v1alpha1/defaults.go [hunk 38]
```diff
@@ -18,13 +18,28 @@ package v1alpha1
...
+		spec.RelistBehavior == "Manual"
+
+	if !isValidRelistBehavior {
```
Comment (nilebox) [inline pkg/apis/servicecatalog/v1alpha1/defaults.go]: Defaults shouldn't handle invalid values, they handle missing values.
Hence, instead of the check above you should check for empty string here.

File: pkg/apis/servicecatalog/v1alpha1/defaults.go (hunk 37 above)
Comment (nilebox) [reply pkg/apis/servicecatalog/v1alpha1/defaults.go]: This code shouldn't be there at all, see below.

File: pkg/apis/servicecatalog/v1alpha1/defaults_test.go (hunk 24 above)
Comment (nilebox) [reply pkg/apis/servicecatalog/v1alpha1/defaults_test.go]: @eriknelson yes, it should be enforced to nil (and checked at the validation stage). similar to union types, where only one branch can be initialized at a time.

File: pkg/apis/servicecatalog/validation/broker.go [hunk 39]
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
+		spec.RelistBehavior == "Manual"
+	if !isValidRelistBehavior {
+		errMsg := fmt.Sprintf("unknown relistBehavior found: %v", spec.RelistBehavior)
+		allErrs = append(
+			allErrs,
+			field.Required(fldPath.Child("relistBehavior"), errMsg),
+		)
+	}
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration && spec.RelistDuration == nil {
```
Comment (nilebox) [inline pkg/apis/servicecatalog/validation/broker.go]: as mentioned in the other comments thread, you should also check that `RelistDuration` is nil when `RelistBehavior != Duration`

File: pkg/apis/servicecatalog/validation/broker_test.go [hunk 40]
```diff
@@ -40,7 +41,9 @@ func TestValidateServiceBroker(t *testing.T) {
 					Name: "test-broker",
 				},
 				Spec: servicecatalog.ServiceBrokerSpec{
-					URL: "http://example.com",
+					URL:            "http://example.com",
+					RelistBehavior: "Duration",
```
Comment (nilebox) [inline pkg/apis/servicecatalog/validation/broker_test.go]: use constant from types (here and below).

File: pkg/apis/servicecatalog/validation/broker_test.go [hunk 41]
```diff
@@ -40,7 +41,9 @@ func TestValidateServiceBroker(t *testing.T) {
...
+					RelistDuration: &metav1.Duration{15 & time.Minute},
```
Comment (nilebox) [inline pkg/apis/servicecatalog/validation/broker_test.go]: should be `*` instead of `&`. here and below.

File: pkg/storage/tpr/storage_interface_test.go [hunk 42]
```diff
@@ -140,7 +141,9 @@ func TestCreateWithNoNamespace(t *testing.T) {
...
+			RelistBehavior: "Duration",
```
Comment (nilebox) [inline pkg/storage/tpr/storage_interface_test.go]: use constant

File: pkg/storage/tpr/storage_interface_test.go [hunk 43]
```diff
@@ -140,7 +141,9 @@ func TestCreateWithNoNamespace(t *testing.T) {
...
+			RelistDuration: &metav1.Duration{15 * time.Minute},
```
Comment (nilebox) [inline pkg/storage/tpr/storage_interface_test.go]: What are we trying to check there by setting those fields? If we want to make sure that serialization and persistence works correctly, you should set values different from default.

File: pkg/apis/servicecatalog/validation/broker.go (hunk 32 above)
Comment (staebler) [reply pkg/apis/servicecatalog/validation/broker.go]: Or use field.NotSupported which takes a []string of valid values.

Comment (staebler) [reply pkg/apis/servicecatalog/validation/broker.go]: Or use field.NotSupported which takes a []string of valid values.

**Comment by eriknelson** [comment]
@nilebox @pmorie think I should have addressed your comments in the last couple commits, also rebased and repaired conflicts. Let me know if I missed something, thank you for the reviews!

File: pkg/controller/controller_broker_test.go [hunk 44]
```diff
@@ -133,12 +138,33 @@ func TestShouldReconcileServiceBroker(t *testing.T) {
...
 			}(),
 			now:       time.Now(),
-			interval:  3 * time.Hour,
 			reconcile: true,
 		},
+		{
+			name: "ready, duration behavior, nil duration",
+			broker: func() *v1alpha1.ServiceBroker {
+				broker := getTestServiceBrokerWithStatus(v1alpha1.ConditionTrue)
+				broker.Spec.RelistBehavior = "Duration"
```
Comment (nilebox) [inline pkg/controller/controller_broker_test.go]: const

File: pkg/apis/servicecatalog/testing/fuzzer.go [hunk 45]
```diff
@@ -192,6 +193,11 @@ func FuzzerFor(t *testing.T, version schema.GroupVersion, src rand.Source) *fuzz
...
+			c.FuzzNoCustom(bs)
+			bs.RelistBehavior = "Duration"
```
Comment (nilebox) [inline pkg/apis/servicecatalog/testing/fuzzer.go]: const

File: pkg/controller/controller_broker_test.go [hunk 46]
```diff
@@ -133,12 +138,33 @@ func TestShouldReconcileServiceBroker(t *testing.T) {
...
+				return broker
+			}(),
+			now:       time.Now(),
+			reconcile: false,
+		},
+		{
+			name: "ready, manual behavior",
+			broker: func() *v1alpha1.ServiceBroker {
+				broker := getTestServiceBrokerWithStatus(v1alpha1.ConditionTrue)
+				broker.Spec.RelistBehavior = "Manual"
```
Comment (nilebox) [inline pkg/controller/controller_broker_test.go]: const

**Comment by eriknelson** [comment]
@nilebox think I got the last of them. For my edification, I had someone suggest it might be preferred to not use the consts from types in tests in case the values change? I'm probably misunderstanding that, but are there any cases where an inline value might be preferred?

**Comment by nilebox** [comment]
@eriknelson if you want to ensure that the constant hasn't changed, I would write a specific test for it. Might be something as trivial as
```
if RelistDuration != "Duration" {
  t.Fail();
}
```

File: pkg/apis/servicecatalog/validation/broker.go [hunk 47]
```diff
@@ -98,13 +98,49 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
...
+		allErrs = append(
+			allErrs,
+			field.Required(fldPath.Child("relistDuration"), "relistDuration must be set if behavior is set to Duration"),
+		)
+	}
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorManual && spec.RelistDuration != nil {
+		allErrs = append(
+			allErrs,
+			field.Required(fldPath.Child("relistDuration"), "relistDuration must not be set if behavior is set to Manual"),
```
Comment (pmorie) [inline pkg/apis/servicecatalog/validation/broker.go]: s/behavior/relistBehavior/

File: pkg/apis/servicecatalog/validation/broker.go (hunk 22 above)
Comment (pmorie) [reply pkg/apis/servicecatalog/validation/broker.go]: Looks like this still needs to be done

**Comment by pmorie** [comment]
LGTM
